File: L1_insertion_sort.py
import random

N = 10

arr = [0] * N

# Function to print the array
def print_array(A):
    print("Array contents:", A)

def insertion_sort(A, n):
	"""Sort a list or numpy array.

	Argument:
	A -- a list or numpy array
	n -- length of A
	"""
	# Traverse the list or array from index 1 to n-1.
	for i in range(1, n):
		key = A[i]

		# Insert A[i] into the sorted subarray a[0:i].
		# Compare stored key with the already sorted values to its left.
		# Move each item one position to the right until we find the 
		# position for the key or fall off the left end of the list or array.
		j = i - 1
		while j >= 0 and A[j] > key:
			A[j + 1] = A[j]
			j -= 1

		# Insert key at the correct position in the list or array.
		A[j + 1] = key
  
def main():

    print("Lab 1 - insertion sort")
    
    for i in range(N):
        arr[i] = random.randint(0, 99)

    print("List before sorting:")
    print_array(arr)
    
    insertion_sort(arr, N)
    
    print("List after sorting:")
    print_array(arr)

if __name__ == "__main__":
    main()

File: L1_insertion_sort.c
#include <stdio.h>
#include <stdlib.h>

#define N 10 

void print_array(int A[], int n) {
    printf("Array contents: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", A[i]);
    }
    printf("\n");
}

void insertion_sort(int A[], int n) {
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = A[i];
        j = i - 1;

        while (j >= 0 && A[j] > key) {
            A[j + 1] = A[j];
            j = j - 1;
        }
        A[j + 1] = key;
    }
}


int main() {
    int arr[N];

    printf("Lab 1 - insertion sort\n");

    // // Set each element of the array to a random value in the range 0-127
    // for (int i = 0; i < N; i++) {
    //     arr[i] = rand() >> 24;  // Right shift by 24 bits to get a value between 0 and 127
    // }

    // Set each element of the array to a random value in the range 0-127
    for (int i = 0; i < N; i++) {
        arr[i] = rand() % 128; 
    }

    printf("List before sorting:\n");
    print_array(arr, N);

    insertion_sort(arr, N);

    printf("List after sorting:\n");
    print_array(arr, N);

    return 0;
}
